File: sdk/typescript/examples/devices-24x7.ts
/**
 * Always-on device monitoring: poll live snapshots on a fixed interval,
 * forever, with simple offline detection and SIGINT-safe shutdown. This is
 * the pattern for 24/7 services that watch a Voltrus server.
 *
 * Environment:
 *   VOLTRUS_BASE_URL            server base URL   (default http://localhost:3000)
 *   VOLTRUS_API_KEY             API key credential
 *   VOLTRUS_USERNAME            local login alternative
 *   VOLTRUS_PASSWORD            local login alternative
 *   VOLTRUS_POLL_INTERVAL_MS    poll interval     (default 2000)
 *
 * Run: npx tsx examples/devices-24x7.ts
 */
import { VoltrusClient } from '../src';

const intervalMs = Number(process.env.VOLTRUS_POLL_INTERVAL_MS ?? 2000);
const MAX_BACKOFF_MS = 30000;

async function connect(): Promise<VoltrusClient> {
  const client = new VoltrusClient({
    basePath: process.env.VOLTRUS_BASE_URL ?? 'http://localhost:3000',
    apiKey: process.env.VOLTRUS_API_KEY,
    timeout: Math.min(intervalMs, 5000),
  });
  if (!process.env.VOLTRUS_API_KEY) {
    const username = process.env.VOLTRUS_USERNAME;
    const password = process.env.VOLTRUS_PASSWORD;
    if (!username || !password) {
      console.error('Set VOLTRUS_API_KEY or VOLTRUS_USERNAME + VOLTRUS_PASSWORD');
      process.exit(1);
    }
    await client.auth.login(username, password);
  }
  return client;
}

async function pollOnce(client: VoltrusClient, pollCount: number): Promise<void> {
  const live = await client.system.getLive();
  const online = live.devices.filter((d) => d.status === 'online');
  const stamp = new Date().toISOString();

  console.log(`\n[${stamp}] poll #${pollCount}: ${online.length}/${live.devices.length} devices online`);
  for (const device of live.devices) {
    const metrics = Object.entries(device.metrics)
      .slice(0, 4)
      .map(([name, value]) => `${name}=${value}`)
      .join('  ');
    const marker = device.status === 'online' ? '+' : 'x';
    console.log(`  [${marker}] ${device.device_id.padEnd(24)} ${metrics}`);
  }
}

async function main(): Promise<void> {
  const client = await connect();
  console.log(`Watching ${process.env.VOLTRUS_BASE_URL ?? 'http://localhost:3000'} every ${intervalMs}ms — Ctrl+C to stop.`);

  let consecutiveFailures = 0;
  let polls = 0;
  let running = true;

  process.on('SIGINT', () => {
    console.log(`\nShutting down after ${polls} polls.`);
    running = false;
  });

  while (running) {
    try {
      polls += 1;
      await pollOnce(client, polls);
      consecutiveFailures = 0;
      await new Promise((resolve) => setTimeout(resolve, intervalMs));
    } catch (error) {
      consecutiveFailures += 1;
      // Linear backoff: the server may be restarting for maintenance; keep
      // trying instead of crashing the watcher.
      const backoff = Math.min(intervalMs * consecutiveFailures, MAX_BACKOFF_MS);
      const message = error instanceof Error ? error.message : String(error);
      console.error(`Poll failed (${consecutiveFailures} in a row): ${message} — retrying in ${backoff}ms`);
      await new Promise((resolve) => setTimeout(resolve, backoff));
    }
  }
}

main().catch((error) => {
  console.error('Watcher failed:', error instanceof Error ? error.message : error);
  process.exit(1);
});

File: sdk/python/examples/devices_24x7.py
"""24/7 device poller: run forever, log device snapshots and live values.

The always-on production pattern: fixed-interval polling, transient errors
logged and retried (the loop never dies on a network blip), clean Ctrl-C.

Env vars:
    VOLTRUS_URL       base URL of the server (default http://localhost:3000)
    VOLTRUS_API_KEY   API key; or set USERNAME + PASSWORD instead
    VOLTRUS_USERNAME  login user (used when VOLTRUS_API_KEY is not set)
    VOLTRUS_PASSWORD  login password
    VOLTRUS_INTERVAL  poll interval in seconds (default 5)
    VOLTRUS_DEVICE    optional device id to filter live values by

Usage: python devices_24x7.py
"""

import os
import signal
import sys
import time

from quickstart import make_client

from voltrus import VoltrusError

_running = True


def _stop(signum: int, frame: object) -> None:
    global _running
    _running = False


def poll_once(client: object, device_id: str) -> None:
    snapshots = client.get_live_snapshots()
    online = sum(1 for s in snapshots if s.status == "online")
    print(
        f"[{time.strftime('%H:%M:%S')}] devices {online}/{len(snapshots)} online"
    )
    for s in snapshots:
        metrics = ", ".join(f"{k}={v:g}" for k, v in sorted(s.metrics.items())[:4])
        print(f"    {s.device_id:<24} {s.status:<8} {metrics}")

    values = client.get_live_data(device_id=device_id)
    warnings = [v for v in values if v.status in ("warning", "critical")]
    for v in warnings:
        print(f"    ALARM-RISK {v.key} = {v.value:g} ({v.status})")


def main() -> None:
    client = make_client()
    interval = int(os.environ.get("VOLTRUS_INTERVAL", "5"))
    device_id = os.environ.get("VOLTRUS_DEVICE", "")

    signal.signal(signal.SIGINT, _stop)
    signal.signal(signal.SIGTERM, _stop)

    print(f"Polling every {interval}s — Ctrl-C to stop.")
    while _running:
        started = time.monotonic()
        try:
            poll_once(client, device_id)
        except VoltrusError as e:
            # Transient API errors must not kill a 24/7 process; log and retry.
            print(f"API error {e.status}: {e.message} — retrying next cycle")
        except OSError as e:
            print(f"Network error: {e} — retrying next cycle")

        delay = interval - (time.monotonic() - started)
        if delay > 0 and _running:
            time.sleep(delay)

    print("Stopped.")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("Stopped.")
    except SystemExit as e:
        sys.exit(e.code)
